Dont error on 0-dim in convolution

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52614 Avoid dispatch overhead in call to mkldnn convolution
* #52613 Include max pool in fusion groups
* #52513 Extend subgraph utils to cover merging a node following a subgraph
* #52512 Remove unused subgraph vmap api
* #51924 Only functionalize ops which we want to include in mkldnn group
* #51923 Add a filter to remove mutation
* **#51922 Dont error on 0-dim in convolution**
* #51921 Add 0-dim support for binary mkldnn ops
* #51736 Handle mkldnn broadcasting in mkldnn fuser
* #51600 Add MKLDNN fuser
* #52786 [NNC] Make NNC sanitize input names
* #51484 Add support for linear in mkldnn fusion
* #51483 Move conv to mkldnn

Differential Revision: [D26696701](https://our.internmc.facebook.com/intern/diff/D26696701)

Commit:
Dont error on 0-dim in convolution

[ghstack-poisoned]

Commit:
Update on "Dont error on 0-dim in convolution"




[ghstack-poisoned]

Krovatkin commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -643,6 +643,15 @@ at::Tensor _convolution(
                           params.output_padding, params.stride, params.dilation,
                           params.groups);
     }
+    if (input_is_mkldnn && weight.is_mkldnn()) {
```
this is under `input.size(0) == 0` ?

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "Dont error on 0-dim in convolution"




[ghstack-poisoned]

Commit:
Update on "Dont error on 0-dim in convolution"




[ghstack-poisoned]

Commit:
Update on "Dont error on 0-dim in convolution"




[ghstack-poisoned]

Commit:
Update on "Dont error on 0-dim in convolution"




[ghstack-poisoned]

Commit:
Update on "Dont error on 0-dim in convolution"




[ghstack-poisoned]

Krovatkin commented on test/test_mkldnn.py:
```diff
@@ -703,6 +703,14 @@ def test_0_dimension_tensor(self):
+        C = 7
+        m = mkldnn_utils.to_mkldnn(torch.nn.Conv2d(C, C, 3))
+        x = torch.randn(0, C, C, 8, dtype=torch.float)
+        out_eager = m(x)
```
hasn't `m` already been converted to mkldnn?
are we testing here `eager` vs `mkldnn` ?

eellison replied:
yea, good catch

Krovatkin reviewed (changes requested):
a bit confused by the test

Commit:
Update on "Dont error on 0-dim in convolution"




[ghstack-poisoned]

Commit:
Update on "Dont error on 0-dim in convolution"




[ghstack-poisoned]

Commit:
Update on "Dont error on 0-dim in convolution"




[ghstack-poisoned]

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "Dont error on 0-dim in convolution"




[ghstack-poisoned]

Commit:
Update on "Dont error on 0-dim in convolution"


Differential Revision: [D26696701](https://our.internmc.facebook.com/intern/diff/D26696701)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/Convolution.cpp:
```diff
@@ -651,6 +651,15 @@ at::Tensor _convolution(
     }
+    if (input_is_mkldnn && weight.is_mkldnn()) {
+      // mkldnn will error on the below 0-dim handling code
+      return empty_mkldnn(
```
If support for empty binary ops is added in the previous PR, why does the code below still error out for mkldnn?
This is wrong, because with empty_mkldnn you are returning non-differentiable output.

eellison replied:
Should we revert this ?

ngimel replied:
Why does it fail in the subsequent code? Is it the view ops? To preserve gradient flow, your output needs to be connected to weight and bias, so all you really need to do is be able to multiply and add with empty input, then the output has the correct history.
